### ISSUE TITLE
Transformer does not work?

### ISSUE THREAD

**Issue by Jey-Cee** [enocean-js/enocean-js#166]
Hi Holger,

i am trying to use enocean-js and later contribute to it. But it isnt easy for me to use it, i have spend a day to understand what it is for what and how to import it as none es6 module.

Now next show stopper is the transformer, it is getting the data-telegram from the parser and the output is exactly the same as the input.
I don't see why it is not working, my code is like the examples i found in the repositories.
Can you help me?

Here my code:
`
const ESP3Parser = require('@enocean-js/serialport-parser').ESP3Parser;
const ESP3Transformer = require('@enocean-js/esp3-transformer').ESP3Transformer;

let SERIAL_PORT = new SerialPort(this.config.serialport, { baudRate: 57600});

const parser = new ESP3Parser();
const transformer = new ESP3Transformer();
SERIAL_PORT.pipe(parser).pipe(transformer);
parser.on('data', async data => { console.log(data); });
transformer.on('data', async data => { console.log(data); });
`

**Comment by Jey-Cee** [comment]
I found this [line](https://github.com/enocean-js/enocean-js/blob/750cfa41dc1f05cecb2a62190d6bf6741640ad0f/packages/node_modules/%40enocean-js/radio-erp1/src/radio-erp1.js#L202), that says the telegram is a ESP3Packet. 
And now, how can i get this telegram decoded?

**Comment by Holger-Will** [comment]
ESP3Packet is "base" class, (almost) every telegram is an ESP3Packet. the transformer tuirns the packets into their concrete type. For data-telegrams that would be RadioERP1. When you have a RadioERP1 you can decode it with `radio.decode(eep)`

so for example:
```
transformer.on('data', async data => { 
  console.log(data.decode('a5-02-03'))
})
```
as you can see for this to work, you need to know the eep of the telegrams you receive. 
you can either get them from the manual of the device or by listening to teach in telegrams (teach in telegrams are RadioERP1 telegrams as well).

```
let known = {}
transformer.on('data', async data => { 
  if(known .hasOwnProperty(data.senderId)){
    // we know this device... decode
   console.log(data.decode(known[data.senderId].eep))
  }
  if(data.teachIn){
    // we have a teach in... store this device info 
    console.log(data.teachInInfo)
    known[data.teachInInfo.senderId ]= {senderId: data.teachInInfo.senderId, eep: data.teachInInfo.eep.toString()})
  }
})
```
(this is untested code, just written of the top of my head... but it should work kind of like this...)

**Comment by Holger-Will** [comment]
this is unless you are in Japan, where everything is not a RadioERP1 but rather a RadioERP2 enoded package... that code path is largely untested because.. well I'm not in Japan. I only have one Japanese device and one Japanese USB stick here for testing :-(
So if you are in Japan we have to wander off into unknown territory ;-)

**Comment by Jey-Cee** [comment]
Now i am really confused. 
If every telegram is an ESP3Packet this `SERIAL_PORT.pipe(parser).pipe(transformer);` will never work. 
So why this is in the examples used?

**Comment by Holger-Will** [comment]
well, let me rephrase that a bit, every telegram like RadioERP1, RadioERP2, Message, CommonCommand, etc. is a ESP3Packet. the parser just knows about ESP3Packets and only emits these. An ESP3Packet consists of a Header and a Payload it does not, however, know anything about the encoding used in the Payload...
The parser hands these ESP3Packets over to the transformer. The transformer looks into the ESP3Packet header and determines the packetType (RadioERP1, RadioERP2, etc) and emits these.
RadioERP1 are still ESP3Packets, but their payload is encoded differently from RadioERP2, etc.

I will post an example here later to make it more clear...

**Comment by Holger-Will** [comment]
`55000707017af-6000027820320-00ffffffff4a00-50`

let's take this packet. what ESP3 tells us is that the first 6 Bytes are the header of the message:

Header: `55000707017a`  

breaking this down...

`55` (this is an enocean packet) `0007` (the `data` part is 7 Bytes long) `07` (the `optionalData` part is also 7 Bytes long) `01` (it is a RadioERP1 Packet) `af` (the checksum of the header)...
from this header, we know that the following 14 Bytes are the `data` and `optionalData` part.

data: `f6000027820320`
optionalData: `00ffffffff4a00`

and finally, the last Byte `50` is the checksum for the body. And this is all ESP3 specifies. But from Byte 4 of the header, we know that we have to look into the RadieERP1 specification to understand the data and optionalData part. To 'convert' an ESP3Packet to a RadioERP1 packet we can either use the 'transformer' or do it by hand:


```
import { RadioERP1 } from '../../radio-erp1/src/radio-erp1.js'
...
parser.on('data', async data => { 
  let tel = RadioERP1.from(data)
  console.log(tel.senderId)
});
```
the adventage of using the transformer is that it already knows about the different packetTypes.

so what does the RadioERP1 spec tell us?

we have our data (`f6000027820320`) and optionaData (`00ffffffff4a00`) looking at the data, RadioERP1 tells us that the first Byte is the telegram type 

telegramType: `f6`

`f6` means it's a 1 Byte communication telegram, so the next Byte is the payload.

payload: `00`

RadioERP1 does NOT tell us anything about the encoding of the payload. For that, we need the EEP specification which is implemented in the transcoder (more on that later)...

the next 4 Bytes is the senderId: `00278203`

and finally, we have the status byte: `20`

and now the optionalData: `00ffffffff4a00`

here we have the repeatCount (`00`) this is used internally by repeaters for example.
the following 4 Bytes is the destinationId: `ffffffff` (this means its a broadcast message)

the next Byte `4a` is the RSSI value that means it's the signal strength in db...
and the last Byte ('00') is the encryption... so this is not encrypted...

this is RadioERP1 know about a packet. So finally to decode the payload of a RadioERP1 packet we need the EEP spec.    
I will post a more detailed description about EEP later...

**Comment by Jey-Cee** [comment]
Thanks now it is getting brighter in the dark. 
I know the enocean specs, i have written the enocen adapter for ioBroker. Ok, this was back in 2017, but i could rember the basics.

For me it is hard to understand how to use the enocean-js packets. 
I think a flow chart showing which packet interacts with which would help to better understand how enocean-js works.

**Comment by Holger-Will** [comment]
that is a goods idea! 
I know the documentation is virtually nonexistent :-(
we are at version 0.0.10 (still very early) and last year I focused mainly on node-red integration...

if I can help you in any way understanding the code please let me know!

**Comment by Holger-Will** [comment]
I'm thinking about getting rid of the transformer completely and just have the parser return the concrete types...

**Comment by Jey-Cee** [comment]
Ok i dont know what are the consequences or how many work it is. But i think less modules are better for overview.

**Comment by Holger-Will** [comment]
it's not much work. I just implemented a parser on top of the web-serial spec and putting the transformer logic into the parser equals to adding 5 lines of code ;-)

**Comment by Holger-Will** [comment]
starting from version `0.0.15` using the transformer is optional. the parser will emit the concrete types already... We still have to update the documentation and examples...

**Comment by Holger-Will** [comment]
so now you can do

```
const parser = new ESP3Parser();
SERIAL_PORT.pipe(parser)
parser.on('data', async data => { console.log(data.senderId); });
```
